### ISSUE TITLE
Processes calling brod:produce_sync sometimes block forever

### ISSUE THREAD

**Issue by Kwisatx** [kafka4beam/brod#189]
Hello again,

We've been investigating issues where our number of process would slowly increase. This always happens after one of the kafka server nodes becomes unavailable (for even a very short time), leading to kafka push errors (NotLeaderForPartition) during the unavailability. Once the node is back, things look like they are back to normal, except that we have the steady increase in blocked processes.

All the blocked processes are stuck in brod_producer:produce line 150, with the following erlang stacktraces:
```
 {{current_stacktrace,[{brod_producer,produce,3,[{file,"src/brod_producer.erl"},{line,150}]},
                       {brod,produce_sync,5,[{file,"src/brod.erl"},{line,340}]},
                       [**our process callstack**] ]},
```
This corresponds to a receive, waiting for #brod_produce_reply with result=brod_produce_req_buffered.

I added some debug to brod, waited for the issue to reappear, and I think I've cornered why they block forever:
The processes call: brod:produce_sync -> brod:produce -> brod_producer:produce -> Pid ! {produce, CallRef, Key, Value} & receive waiting for #brod_produce_reply with result=brod_produce_req_buffered.

However, in brod_producer_buffer:maybe_buffer, if BufferCount >= BufferLimit we will never call reply_buffered(Req), which means that the process will never receive brod_produce_req_buffered.

I presume the temporary push errors lead to the buffer filling up, which is why we don't usually see the issue, but I'm not sure :p.

Does my analysis hold up? Or am I misunderstanding what is happening?

Edit: Reproduced with brod 2.3.1 (originally we found the bug with 2.2.12) and kafka 0.10.1.1

**Comment by zmstone** [comment]
Hi @Kwisatx 
I assume you are using the latest version?
I don't think a caller can be blocked 'forever'.

There are two queues in `brod_producer_buffer`, buffer queue and pending queue.
The caller is indeed blocked when buffer is full,
but the reply should be sent when the call is taken from pending queue into buffer queue.
https://github.com/klarna/brod/blob/master/src/brod_producer_buffer.erl#L210

**Comment by Kwisatx** [comment]
Yep, latest version.

Hmm, I added debug to reply_buffered which never got triggered if we went through the full buffer case.

From what I can tell line 210 calls maybe_buffer with an ?EMPTY_QUEUE, and maybe_buffer(#buf{ pending      = ?EMPTY_QUEUE } = Buf) doesn't call reply_buffered, so no reply from there.

**Comment by Kwisatx** [comment]
Ah, my bad, one's in pending the other in buffer, hmm.

**Comment by zmstone** [comment]
Just sent an enhancement PR which should give better latency when the buffer is often smaller than batch.
But I still can't think of a scenario where the reply can be missing.

**Comment by Kwisatx** [comment]
Hi again @zmstone,

I've got some more information after adding some debug, but I'm still analyzing everything that's happening :p.
From what I can tell the producer process is stuck with a message sent on the wire which is somehow never acked. This actually happens before the buffer is full however. We only noticed it once the buffer was full because most of our processes were blocking waiting for the buffer reply, but there were actually a number of processes waiting for the ack reply as well. 

Basically the sequence if the buffer is full is:
- the producer calls brod_producer_buffer:add but the buffer is full
- the producer then calls maybe_produce which calls brod_producer_buffer:maybe_send
- however take_reqs_to_send sees that OnWireCount == 1 == OnWireLimit  and returns the buffer unchanged and doesn't send anything.

If the buffer isn't full yet it's the same general idea.

This continued for a bit more than an hour and a half, and then this time the situation resolved itself. I didn't put any debug in prod_producer:handle_info({msg, Pid, CorrId, #kpro_ProduceResponse{} = R}, more's the pity, so I'm not sure what happened. Especially since none of my debug in take_reqs_to_send or maybe_buffer was tripped, which would seem to mean that the whole buffer was just given up on some how.

Can you think of any reason there could be for an ack "getting stuck"? Or any good ideas for where to add some debug for next time ;)?

I haven't tried the PR yet, as I wanted to reproduce the issue with a bit more debug.

Thanks for your help :)

**Comment by zmstone** [comment]
could you paste your producer config here? if there is any.

**Comment by zmstone** [comment]
and client config too

**Comment by Kwisatx** [comment]
`ok = brod:start_client(HostPorts, brod_client_producer, [{auto_start_producers, true}, {default_producer_config, [{ack_timeout, ?KAFKA_TIMEOUT}, {max_retries, ?KAFKA_RETRIES}]}]),`

with 
`-define(KAFKA_TIMEOUT, 1000).          %% Time to wait for ack after pushing to kafka, 1s is default`
`-define(KAFKA_RETRIES, 3).             %% Number of retries when trying to write to kafka, 3 is default`

**Comment by zmstone** [comment]
@Kwisatx Thanks! Nothing seem wrong in the configs.

Sorry, I still can't think of a reason, have to ask more questions instead :)

You mentioned `NotLeaderForPartition` error code,
I assume it was in the `Error in produce response` error log?
https://github.com/klarna/brod/blob/2.3.2/src/brod_producer.erl#L274
Did you notice any `brod_sock` crash logs ?

> This continued for a bit more than an hour and a half

Was beam process idle (or at least not too busy) ? --- to exclude dead loops.
How did you tell the time span? from the timestamps in debug logs you added ?

One possible cause could be that the correlation ID for a response was discarded.
for example at this line: https://github.com/klarna/brod/blob/2.3.2/src/brod_producer.erl#L286
Maybe you can try add logs there (I'll add logs in PR #190 too).

**Comment by Kwisatx** [comment]
We saw the NotLeaderForPartition in the `Error in produce response` error logs, as well as when we reach the max_retries and the process crashes:

> 2017-03-29 14:21:30.050 UTC [error] <0.16254.1> Error in produce response
> Topic: presences
> Partition: 7
> Offset: -1
> Error: 'NotLeaderForPartition'
> 2017-03-29 14:21:30.050 UTC [error] <0.16254.1> gen_server <0.16254.1> terminated with reason: {reached_max_retries,{produce_response_error,<<"presences">>,7,-1,'NotLeaderForPartition'}}

There are also a number of RequestTimedOut errors around as well.

I don't see any brod_sock crashes in the logs.

I'll investigate the beam process more next time, one thing we can say is that the server itself wasn't too busy during the issues.

I determined the timespan by on one hand the number of processes that we monitor, and also by looking at the timestamps of the debug I added. I followed one of the producer processes that was stuck with an OnWireCount of one.

Welp, I'm adding different debug, will see what information I can glean this time :p

**Comment by zmstone** [comment]
So the `brod_producer` process terminated
The callers got stuck after a restart I suppose.
> I followed one of the producer processes that was stuck with an OnWireCount of one.

did you do `sys:get_state` or ?

**Comment by Kwisatx** [comment]
The thing is that the processes blocking happen after the produce errors (and brod_producer restarts), and thus with a different brod_producer (that is still alive as it receives new calls, and calls protocol_buffer). The issue is that new callers are getting stuck.
Initially we thought along the same lines, except that after the kafka issue is resolved, the number of blocked processes just keeps rising.

I didn't call `sys:get_state`, I just added a lot of debug in the calls (with the pid in the log). I also looked at the process_info, but nothing stood out. I'll do that next time we see the issue.

**Comment by zmstone** [comment]
@Kwisatx Thanks!
It would be more clear if we can see the process state and process info (current function etc) of the `brod_producer` process that got stuck.
You can call `brod:get_producer` API to get its pid.

**Comment by Kwisatx** [comment]
Hi @zmstone , great news, the problem's back :p!

You can find the results of process_info and sys:getstate here : 
[blocked_producer.txt](https://github.com/klarna/brod/files/890614/blocked_producer.txt)

I'm currently analyzing the logs, I'll get back to you with what I can gather.

**Comment by Kwisatx** [comment]
Huh, I suppose it's worth noting that if I do a sys:getstate on the client (brod_client_producer in my case), the metadata socket is still considered dead: {dead_since,{1491,225330,720060},{shutdown,tcp_closed}}.
I don't know if that's interesting information ^^.

Otherwise I don't really know what to make of the logs, there just stop being acks from what I can see: 
[brod_producer_bufffer_logs.txt](https://github.com/klarna/brod/files/890750/brod_producer_bufffer_logs.txt)

**Comment by zmstone** [comment]
@Kwisatx Thank you!
Metadata socket will be closed by kafka after idle for 10 minutes (default).
It would be interesting though to know the process info of the payload socket
(Pid <3058.199.0> in blocked_producer.txt).

From the printed state, I can see the first in-flight request has a retry counter 1.
Looking into the logs now.

**Comment by zmstone** [comment]
@Kwisatx I think I have found the issue.
When the socket process is DOWN after a request has been sent,
the in-flight request are not put back to buffer (i.e. should have called schedule_retry/1 not /2).

Working on a fix.

**Comment by zmstone** [comment]
no, not really, it is schedule_retry/2 called when socket is DOWN.
I need to get some sleep, and look again tomorrow.

**Comment by zmstone** [comment]
alright, I believe this is what happened:
```
1. send produce request (onwire = 1)
2. received not leader error (nack_all, onwire = 0)
3. schedule retry (onwire = 0)
4. produce request triggered by a new produce call (onwire = 1, retry_count = 1)
5. scheduled retry send_after loop-back message received, AND discovered a new leader
6. produce response for the request sent on step 4 is discarded due to socket pid mismatch
```
The bug should have been fixed in 2.3.3 where a produce request can not be sent
when the retry timer is started.
@Kwisatx Please try 2.3.3
Nonetheless, for better safety, I'll work on a fix to `nack_all` when a new leader is discovered.

EDIT: typo on version number fixed.

**Comment by zmstone** [comment]
@Kwisatx Sorry, 2.3.3 should have the same issue, only less chances to get triggered. To be fixed in 2.3.4.
Thanks a lot, again, for your awesome investigation.

**Comment by Kwisatx** [comment]
@zmstone: sounds good! Thanks for the hard work :).
I'll put the new version in test as soon as it's released.
